application/Common/Shared/Utility/GlobalHook.cs

```diff
@@ -0,0 +1,154 @@
+        private static extern bool capture([MarshalAs(UnmanagedType.U4)] uint type);
+
+        [DllImport(@"HookLibrary64", CallingConvention = CallingConvention.Cdecl)]
+        private static extern void stopCapture([MarshalAs(UnmanagedType.U4)] uint type);
+        #endregion
+
+        #region Public functions
+
+        /// <summary>
+        ///     Add a listener to multiple message types.
+        /// </summary>
+        /// <param name="callback">The <see cref="RetrieveMessageCallBack" /> function to be called on messages.</param>
+        /// <param name="types">An array of the types to listen for.</param>
+        public static void addListener(RetrieveMessageCallBack callback, params NativeMethods.MessageType[] types)
+        {
+            foreach (var type in types)
+            {
+                if (!capture((uint) type))
+                {
+                    throw new NotSupportedException(
+                        $"GlobalHook currently does not support this message type ({type})");
+                }
+            }
+
+            foreach (var type in types)
+            {
+                if (!listeners.ContainsKey(type))
+                {
+                    listeners.Add(type, new List<RetrieveMessageCallBack>());
+                }
+
+                listeners[type].Add(callback);
+            }
+        }
+
+        /// <summary>
+        ///     Remove a listener from multiple message types.
+        /// </summary>
+        /// <param name="callback">The <see cref="RetrieveMessageCallBack" /> to remove.</param>
+        /// <param name="types">An array of the types to remove the callback for.</param>
+        public static void RemoveListener(RetrieveMessageCallBack callback, params NativeMethods.MessageType[] types)
+        {
+            foreach (var type in types)
+            {
+                if (listeners.ContainsKey(type))
+                {
+                    listeners[type].Remove(callback);
+                    if (listeners[type].Count == 0)
+                    {
+                        listeners.Remove(type);
+                        stopCapture((uint)type);
+                    }
+                }
+            }
+        }
+
+        #endregion
+
+        #region Private functions
+
+        private static void Trigger(NativeMethods.MessageType type, HookMessage message)
+        {
+            if (listeners.ContainsKey(type))
+            {
+                foreach (var cb in listeners[type])
+                {
+                    cb(message);
+                }
+            }
+        }
+
+        private static void Hook()
+        {
+            if (!isActive)
+            {
+                SetHook(cppCallback);
+                isActive = true;
+            }
+        }
+
+        private static void UnHook()
+        {
+            if (isActive)
+            {
+                RemoveHook();
+                isActive = false;
+            }
+        }
```

Comment:
If you end up setting `isActive` manually anyway, I would prefer undoing the changes to `SetAndDispatch` and just writing the setter manually

```
public static bool IsActive
{
    get => isActive;
    set
    {
        if (value)
        {
            Hook();
        }
        else
        {
            UnHook();
        }
    }
}
```